delivery-service/ds/src/main.rs

```diff
@@ -0,0 +1,226 @@
+//! environment.
+//!
+//! Because the infrastructure description doesn't give a lot of guidelines on
+//! the design of the DS we take a couple of deliberate design decisions here:
+//! * The DS does not know about groups.
+//! * Clients have to send a list of clients (group members) along with each
+//!   message for the DS to know where to send the message.
+//! * The DS stores and delivers key packages.
+//!
+//! This is a very basic delivery service that allows to register clients and
+//! send messages to MLS groups.
+//! Note that there are a lot of limitations to this service:
+//! * No persistence layer such that all information gets lost when the process
+//!   shuts down.
+//! * No authentication for clients.
+//! * Key packages can't be updated, changed or deleted at the moment.
+//! * Messages lost in transit are gone.
+//!
+//! **⚠️ DON'T EXPECT ANY SECURITY OR PRIVACY FROM THIS!**
+//!
+//! The server always listens on localhost and should be run behind a TLS server
+//! if accessible on the public internet.
+//!
+//! The DS returns a list of messages queued for the client in all groups they
+//! are part of.
+
+use actix_web::{
+    body::Body, get, post, web, web::Bytes, web::Payload, App, HttpRequest, HttpServer, Responder,
+};
+use clap::App as ClapApp;
+use futures_util::StreamExt;
+use std::collections::HashMap;
+use std::ops::{Deref, DerefMut};
+use std::sync::Mutex;
+
+use ds_lib::*;
+use openmls::prelude::*;
+
+#[cfg(test)]
+mod test;
+
+/// The DS state.
+/// It holds a list of clients and their information.
+#[derive(Default, Debug)]
+pub struct DsData {
+    // (ClientName, ClientInfo)
+    clients: HashMap<String, ClientInfo>,
+}
+
+// === API ===
+
+/// Registering a new client takes a serialised `ClientInfo` object and returns
+/// a simple "Welcome {client name}" on success.
+/// An HTTP conflict (409) is returned if a client with this name exists already.
+#[post("/clients/register")]
+async fn register_client(mut body: Payload, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    let mut bytes = web::BytesMut::new();
+    while let Some(item) = body.next().await {
+        bytes.extend_from_slice(&item.unwrap());
+    }
+    let info = ClientInfo::decode(&mut Cursor::new(&bytes)).unwrap();
+
+    let mut data = data.lock().unwrap();
+    let old = data.clients.insert(info.client_name.clone(), info.clone());
+    if old.is_some() {
+        return actix_web::HttpResponse::Conflict().finish();
+    }
+
+    actix_web::HttpResponse::Ok().body(format!("Welcome {}!\n", info.client_name))
+}
+
+/// Returns a list of clients.
+/// This is informational only right now and a simple string with all registered
+/// client names.
+#[get("/clients/list")]
+async fn list_clients(_req: HttpRequest, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    let data = data.lock().unwrap();
+    actix_web::HttpResponse::Ok().body(format!(
+        "I know these clients {:?}!\n",
```

Comment:
Wouldn't it make more sense to output some actual JSON? This is harder to parse.

Reply:
This is really a left over from how this was done before and only informational really. But sure, we can return JSON.